id-verifier/frontend/src/wallet-connection.tsx

```diff
@@ -0,0 +1,162 @@
+    disconnect?(): Promise<void>;
+
+    /**
+     * @param account string when account is changed, undefined when disconnected
+     */
+    protected onAccountChanged(account: string | undefined) {
+        this.emit('accountChanged', account);
+    }
+}
+
+export class BrowserWalletProvider extends WalletProvider {
+    async connect(): Promise<string | undefined> {
+        const provider = await detectConcordiumProvider();
+
+        provider.on('accountChanged', (account) => super.onAccountChanged(account));
+        provider.on('accountDisconnected', async () =>
+            super.onAccountChanged((await provider.getMostRecentlySelectedAccount()) ?? undefined)
+        );
+
+        // Check if you are already connected
+        const selectedAccount = await provider.getMostRecentlySelectedAccount();
+        return selectedAccount ?? provider.connect();
+    }
+
+    async requestIdProof(statement: IdStatement, challenge: string): Promise<IdProofOutput> {
+        const provider = await detectConcordiumProvider();
+        const account = await provider.connect();
+
+        if (!account) {
+            throw new Error('No account available');
+        }
+
+        return provider.requestIdProof(account, statement, challenge);
+    }
+}
+
+const ID_METHOD = 'proof_of_identity';
+
+export class WalletConnectProvider extends WalletProvider {
+    private account: string | undefined;
+    private topic: string | undefined;
+
+    constructor(private client: SignClient) {
+        super();
+
+        this.client.on('session_update', ({ params }) => {
+            this.account = this.getAccount(params.namespaces);
+            super.onAccountChanged(this.account);
+        });
+
+        this.client.on('session_delete', () => {
+            this.account = undefined;
+            this.topic = undefined;
+
+            super.onAccountChanged(this.account);
+        });
+    }
+
+    static async create() {
+        const client = await SignClient.init(walletConnectOpts);
+        return new WalletConnectProvider(client);
+    }
+
+    async connect(): Promise<string | undefined> {
+        try {
+            const { uri, approval } = await this.client.connect({
+                requiredNamespaces: {
+                    [WALLET_CONNECT_SESSION_NAMESPACE]: {
+                        methods: [ID_METHOD],
+                        chains: [CHAIN_ID],
+                        events: ['accounts_changed'],
+                    },
+                },
+            });
+            // Open QRCode modal if a URI was returned (i.e. we're not connecting an existing pairing).
+            if (uri) {
+                QRCodeModal.open(uri, undefined);
+            }
+            // Await session en approval from the wallet.
+            const session = await approval();
+
+            this.account = this.getAccount(session.namespaces);
+            this.topic = session.topic;
+
+            return this.account;
```

Comment:
Can you explain why we ask for approval also if uri was returned?

In the example they have here https://docs.walletconnect.com/2.0/javascript/sign/dapp-usage they do
```
  // Open QRCode modal if a URI was returned (i.e. we're not connecting an existing pairing).
  if (uri) {
    web3Modal.openModal({ uri });
    // Await session approval from the wallet.
    const session = await approval();
    // Handle the returned session (e.g. update UI to "connected" state).
    // * You will need to create this function *
    onSessionConnect(session);
    // Close the QRCode modal in case it was open.
    web3Modal.closeModal();
  }
```

Reply:
Yes, if we're connecting to an existing pairing we'll already have the account stored in memory (I assume), so maybe this is not necessary.

> I have one question related to session handling. I don't understand why we do it differently than the example in their official documentation.
> 
> But I also don't have enough details to understand if it matters.

The session basically contains information about the wallet connection, and the only information we really care about, is which account is connected, to be able to relay this back to the wallet when requesting an ID proof.

In their documentation, they save everything and extract (I assume) what they need elsewhere. I find the solution done here (as in this repository) pretty much the same, just a simpler version.